crates/core/src/table/file_format_options.rs

```diff
@@ -0,0 +1,308 @@
+use async_trait::async_trait;
+use datafusion::catalog::Session;
+use object_store::path::Path;
+use parquet::basic::Compression;
+use parquet::file::properties::{WriterProperties, WriterPropertiesBuilder};
+use parquet::schema::types::ColumnPath;
+use std::sync::Arc;
+use tracing::info;
+
+#[cfg(not(feature = "datafusion"))]
+#[derive(Clone, Default, Debug, PartialEq)]
+pub struct TableOptions {}
+
+// Top level trait for file format options used by a DeltaTable
+pub trait FileFormatOptions: Send + Sync + std::fmt::Debug + 'static {
+    fn table_options(&self) -> TableOptions;
+
+    fn writer_properties_factory(&self) -> Arc<dyn WriterPropertiesFactory>;
+
+    fn update_session(&self, session: &dyn Session) -> DeltaResult<()> {
+        // Default implementation does nothing
+        Ok(())
+    }
+}
+
+/// Convenience alias for file format options reference used across the codebase
+pub type FileFormatRef = Arc<dyn FileFormatOptions>;
+
+#[derive(Clone, Debug, Default)]
+pub struct SimpleFileFormatOptions {
+    table_options: TableOptions,
+}
+
+impl SimpleFileFormatOptions {
+    pub fn new(table_options: TableOptions) -> Self {
+        Self { table_options }
+    }
+}
+
+#[cfg(feature = "datafusion")]
+impl FileFormatOptions for SimpleFileFormatOptions {
+    fn table_options(&self) -> TableOptions {
+        self.table_options.clone()
+    }
+
+    fn writer_properties_factory(&self) -> Arc<dyn WriterPropertiesFactory> {
+        build_writer_properties_factory_tpo(&Some(self.table_options.parquet.clone())).unwrap()
+    }
+}
+
+#[cfg(feature = "datafusion")]
+pub fn build_writer_properties_factory_ffo(
+    file_format_options: Option<FileFormatRef>,
+) -> Option<Arc<dyn WriterPropertiesFactory>> {
+    file_format_options.map(|ffo| ffo.writer_properties_factory())
+}
+
+#[cfg(feature = "datafusion")]
+pub fn build_writer_properties_factory_or_default_ffo(
+    file_format_options: Option<FileFormatRef>,
+) -> Arc<dyn WriterPropertiesFactory> {
+    build_writer_properties_factory_ffo(file_format_options)
+        .unwrap_or_else(|| build_writer_properties_factory_default())
+}
+
+#[cfg(feature = "datafusion")]
+pub fn to_table_parquet_options_from_ffo(
+    file_format_options: Option<&FileFormatRef>,
+) -> Option<TableParquetOptions> {
+    file_format_options.map(|ffo| ffo.table_options().parquet)
+}
+
+#[cfg(feature = "datafusion")]
+pub fn state_with_file_format_options(
+    state: SessionState,
+    file_format_options: Option<&FileFormatRef>,
+) -> DeltaResult<SessionState> {
+    if let Some(ffo) = file_format_options {
+        ffo.update_session(&state)?;
+    }
+    // Convert file format options to parquet options and delegate to the common implementation
+    let parquet_options = to_table_parquet_options_from_ffo(file_format_options);
+    Ok(state_with_parquet_options(state, parquet_options.as_ref()))
+}
+
+#[cfg(feature = "datafusion")]
+pub fn state_with_parquet_options(
```

Comment:
I still think we should avoid setting table specific options on the state if possible (registering encryption factory is different as the factory also needs to be specified by id in the table options). I will take a look at the cases where this is used and see if we can remove the need for this.